Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -1924,96 +2000,158 @@ public partial class ModuleResolver {
-          if (range.UpperBound != null && ConstantFolder.TryFoldInteger(range.UpperBound) is not null and var hi) {
-            UpdateBounds(null, hi);
+          // Give an info message saying which type was selected--unless the user requested
+          // one particular native type, in which case that must have been the one picked.
+          if (nativeTypeChoices is { Count: 1 }) {
+            Contract.Assert(dd.NativeType == nativeTypeChoices[0]);
+            if (dd.TargetTypeCoversAllBitPatterns) {
+              reporter.Info(MessageSource.Resolver, dd.tok,
+                $"newtype {dd.Name} is target-complete for {{:nativeType \"{dd.NativeType.Name}\"}}");
+            }
+          } else {
+            var targetComplete = dd.TargetTypeCoversAllBitPatterns ? "target-complete " : "";
+            reporter.Info(MessageSource.Resolver, dd.tok,
+              $"newtype {dd.Name} resolves as {targetComplete}{{:nativeType \"{dd.NativeType.Name}\"}} (detected range: {detectedRange})");
           }
-        } else if (bound is ComprehensionExpr.ExactBoundedPool exact && ConstantFolder.TryFoldInteger(exact.E) is not null and var value) {
-          UpdateBounds(value, value + 1);
+
+          return;
         }
       }
+      // Among the choices available to us, we did not find a native type that is big enough and supported by the compiler.
+      if (nativeTypeChoices != null) {
+        reporter.Error(MessageSource.Resolver, dd,
+          "None of the types given in :nativeType arguments is supported by the current compilation target. Try supplying others.");
+      } else if (hasNativeTypePreference == true) {
+        reporter.Error(MessageSource.Resolver, dd,
+          "Dafny's heuristics cannot find a compatible native type. " +
+          "Hint: try writing a newtype constraint of the form 'i: int | lowerBound <= i < upperBound && (...any additional constraints...)'.");
+      }
+    }
 
-      var emptyRange = lowBound != null && highBound != null && highBound <= lowBound;
-      foreach (var nativeT in nativeTypeChoices ?? NativeTypes) {
-        bool lowerOk = emptyRange || (lowBound != null && nativeT.LowerBound <= lowBound);
-        bool upperOk = emptyRange || (highBound != null && nativeT.UpperBound >= highBound);
-        if (lowerOk && upperOk) {
-          bigEnoughNativeTypes.Add(nativeT);
-        } else if (nativeTypeChoices != null) {
+    /// <summary>
+    /// Returns a list of (n, b) pairs, where "n" is a big enough native type to hold "dd" and "b" says whether or not
+    /// all bit patterns of "n" are possible values for "dd".
+    /// Returns null if a failure is detected and reported.
+    /// </summary>
+    [CanBeNull]
+    private List<(NativeType, bool coversEntireIntegerRange)> FigureOutNativeTypeForBitvectorNewtype(NewtypeDecl dd,
+      BitvectorType bitvectorAncestorType, bool noFurtherConstraints,
+      List<NativeType> nativeTypesUnderConsideration, bool reportErrorIfTUCDoesNotFit) {
```

Comment:
You could replace `List<NativeType> nativeTypesUnderConsideration, bool reportErrorIfTUCDoesNotFit` with `List<NativeType> nativeTypesUserOverride` and use the field `NativeTypes` in this method. Same for the next method.

Reply:
`FigureOutNativeTypeForBitvectorNewtype(NewtypeDecl dd, BitvectorType bitvectorAncestorType, bool noFurtherConstraints, List<NativeType> nativeTypesUnderConsideration, bool reportErrorIfTUCDoesNotFit)`

becomes 

```dafny
FigureOutNativeTypeForBitvectorNewtype(NewtypeDecl dd, BitvectorType bitvectorAncestorType, bool noFurtherConstraints, List<NativeType>? nativeTypesUserOverride)

var nativeTypesUnderConsideration = nativeTypesUserOverride ?? NativeTypes;
var reportErrorIfTUCDoesNotFit = nativeTypesUserOverride != null;
```